### PR TITLE
A few more GDP examples

### PR DESCRIPTION
**Pull request by sadavis1**
## Summary/Motivation:
Early in the summer, I worked on a few GDP example problems, the first of which was merged in #2838 . This PR contains the other two GDP example problems from Sawaya (2006) that I implemented. The code solves simple geometric problems and draws pretty pictures.

## Changes proposed in this PR:
- Add "Circles" and "Farm Layout" example code and instance data
- Add matplotlib drawing code for each

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: examples/gdp/farm_layout/farm_layout.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+Farm layout example from Sawaya (2006). The goal is to determine optimal placements and dimensions for farm
+plots of specified areas to minimize the perimeter of a minimal enclosing fence. This is a GDP problem with 
+some hyperbolic constraints to establish consistency of areas with length and width. The FLay05 and FLay06 
+instances may take some time to solve; the others should be fast. Note that the Sawaya paper contains a 
+little bit of nonclarity: it references "height" variables which do not exist - we use "length" for the x-axis 
+and "width" on the y-axis, and it also is unclear on the way the coordinates define the rectangles; we have
+decided that they are on the bottom-left and adapted the disjunction constraints to match.
+"""
+
+from __future__ import division
```
Comment (emma58) [inline examples/gdp/farm_layout/farm_layout.py:11]: Same here--you don't need this import

File: examples/gdp/farm_layout/farm_layout.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    m.perim = Objective(
+        expr=2 * (m.Length + m.Width), doc="Perimeter of the farmland configuration"
+    )
+
+    m.plot_x = Var(
+        m.plots, bounds=(0, m.overall_length_ub), doc="x-coordinate of plot center"
+    )
+    m.plot_y = Var(
+        m.plots, bounds=(0, m.overall_width_ub), doc="y-coordinate of plot center"
+    )
```
Comment (emma58) [inline examples/gdp/farm_layout/farm_layout.py:136]: Are these `doc` arguments right? I thought x and y were bottom lefthand corner in this implementation?

**Review by emma58 — COMMENTED** [review]
This looks good! Just a few nits.

**Comment by bernalde** [comment]
This might be a more involved discussion than this PR, but where should we draw the line between a set of examples, tests, and supported models? In particular, for GDP, we have GDPLib https://github.com/SECQUOIA/gdplib, which could host these two examples. What's the motivation for adding them to the main Pyomo repository?

**Comment by blnicho** [comment]
@bernalde I think this is a great discussion to have outside of this PR. We are inconsistent with how examples are written and tested and also with where they live (both within and outside of the Pyomo repo). I'll open a PEP to start a more comprehensive discussion on this but I don't think this PR needs to wait for it to be resolved.

File: examples/gdp/circles/circles.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
+"""
+The "circles" GDP example problem originating in Lee and Grossman (2000). The 
+goal is to choose a point to minimize a convex quadratic function over a set of 
+disjoint hyperspheres.
+"""
+
+from __future__ import division
```
[the remark this answers is not in the document]
Comment (sadavis1) [reply examples/gdp/circles/circles.py]: Sure

File: examples/gdp/circles/circles.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    Param,
+    RangeSet,
+    Var,
+    Constraint,
+    value,
+)
+
+# Circles2D3 is the original example described in the papers. Ruiz and Grossman
+# (2012) discuss two more example instances, Circles2D36 and Circles3D36, but I
+# couldn't find data for them, so I made a couple.
```
[the remark this answers is not in the document]
Comment (sadavis1) [reply examples/gdp/circles/circles.py]: My initials spell out "SAD" which might confuse the future reader for a minute, so I added an "S. Davis" instead

File: examples/gdp/farm_layout/farm_layout.py (hunk 2 above)
Comment (sadavis1) [reply examples/gdp/farm_layout/farm_layout.py]: Oops, you're right. I had to change that at some point.

**Comment by sadavis1** [comment]
I've made the fixes requested, but I can't answer the broader question about where examples should live and how or if they should be tested. These aren't tested for now, since none of their neighbors were.

**Comment by blnicho** [comment]
@sadavis1 I cancelled the tests to free up GHA resources while we're finalizing things for the release. I'll relaunch them when we're done.

**Review by emma58 — APPROVED** [review]

**Pull request merged**